xarray/core/dataset.py

```diff
@@ -2179,5 +2179,77 @@ def real(self):
         return self._unary_op(lambda x: x.imag, keep_attrs=True)(self)
 
+    def get_variables_by_attributes(self, **kwargs):
+        """Returns a ``Dataset`` with variables that match specific conditions.
+        Can pass in ``key=value ``or ``key=callable``.  Variables are returned
+        that contain all of the matches or callable returns True.  If using a
+        callable note that it should accept a single parameter only,
+        the attribute value.  None is given as the attribute value when the
+        attribute does not exist on the variable.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import pandas as pd
+        >>> import xarray as xr
+        >>> temp = 15 + 8 * np.random.randn(2, 2, 3)
+        >>> precip = 10 * np.random.rand(2, 2, 3)
+        >>> lon = [[-99.83, -99.32], [-99.79, -99.23]]
+        >>> lat = [[42.25, 42.21], [42.63, 42.59]]
+
+        >>> ds = xr.Dataset({'temperature': (['x', 'y', 'time'],  temp, dict(standard_name='air_potential_temperature')),
+        ...                  'precipitation': (['x', 'y', 'time'], precip, dict(standard_name='convective_precipitation_flux'))},
+        ...                 coords={'lon': (['x', 'y'], lon, dict(axis='X')),
+        ...                         'lat': (['x', 'y'], lat, dict(axis='Y')),
+        ...                         'time': pd.date_range('2014-09-06', periods=3),
+        ...                         'reference_time': pd.Timestamp('2014-09-05')})
+        >>> # Get variables with matching "standard_name" attribute
+        >>> ds.get_variables_by_attributes(standard_name='convective_precipitation_flux')
+        <xarray.Dataset>
+        Dimensions:         (time: 3, x: 2, y: 2)
+        Coordinates:
+        * x               (x) int64 0 1
+        * time            (time) datetime64[ns] 2014-09-06 2014-09-07 2014-09-08
+            lat             (x, y) float64 42.25 42.21 42.63 42.59
+        * y               (y) int64 0 1
+            reference_time  datetime64[ns] 2014-09-05
+            lon             (x, y) float64 -99.83 -99.32 -99.79 -99.23
+        Data variables:
+            precipitation   (x, y, time) float64 4.178 2.307 6.041 6.046 0.06648 ...
+        >>> ds.get_variables_by_attributes(standard_name='air_potential_temperature')
+        >>> # Get variables with any axis attribute.
+        >>> ds.get_variables_by_attributes(axis=lambda v: v in ['X', 'Y', 'Z', 'T'])
+        <xarray.Dataset>
+        Dimensions:         (x: 2, y: 2)
+        Coordinates:
+            lat             (x, y) float64 42.25 42.21 42.63 42.59
+            lon             (x, y) float64 -99.83 -99.32 -99.79 -99.23
+        * x               (x) int64 0 1
+        * y               (y) int64 0 1
+            reference_time  datetime64[ns] 2014-09-05
+        Data variables:
+            *empty*
+
+        """
+        vs = []
+
+        has_value_flag  = False
+        for vname, data_array in self.items():
```

Comment:
Can we rewrite this loop not to use `break` for control flow? Something like this is much more comprehensible to me:
```
selection = []
for var_name, variable in self.data_vars.items():
    for attr_name, pattern in kwargs.item():
        attr_value = array.attrs.get(attr_name)
        if (callable(pattern) and pattern(attr_value)
                or attr_value == pattern):
            selection.append(var_name)
```